src/pieces.rs

```diff
@@ -1,86 +1,51 @@
-        let tower = Tower {bottom: bottom, mid: mid, top:top };
-        match tower.is_valid() {
-            true => Ok(tower),
-            false => Err("Invalid tower")
+    /// Returns the top most piece and a tower that has its top piece removed
+    /// Returns Err if the tower is empty
+    /// This function does not modify the original tower
+    pub fn lift_piece(&self) -> Result<(Tower, Piece<'a>), &'static str> {
+        use pieces::Tower::*;
+        match *self {
+            Empty => Err("Cannot lift a piece off an empty tower!"),
+            Single(bottom) => Ok((Empty, bottom)),
+            Double(bottom, middle) => Ok((Single(bottom), middle)),
+            Triple(bottom, middle, top) => Ok((Double(bottom, middle), top))
         }
     }
 
-    pub fn get(&self, position: TowerHeight) -> Option<Piece<'a>> {
-        use pieces::TowerHeight::*;
-        match position {
-            Top => self.top,
-            Middle => self.mid,
-            Bottom => self.bottom,
-            Empty => None,
-        }
-    }
-    // Set the given position to the given piece
-    // Panics on invalid tower setting or when trying to set a piece to the Empty height
-    fn set(&mut self, piece: Option<Piece<'a>>, position: TowerHeight) {
-        use pieces::TowerHeight::*;
-        match position {
-            Top => self.top = piece,
-            Middle => self.mid = piece,
-            Bottom => self.bottom = piece,
-            Empty => panic!("Cannot set a piece at TowerHeight::Empty")
-        }
-
-        if !self.is_valid() {
-            panic!("Attempt to set Tower to an invalid state: {:?}", self);
-        }
-    }
-
-    /// Removes and returns the top most piece from the tower
-    /// Panics if the tower is empty
-    pub fn pop(&mut self) -> Piece<'a> {
-        let height = self.height();
-
-        if height == TowerHeight::Empty {
-            panic!("Cannot pop an empty tower!")
-        }
-        // This unwrap is safe because the tower is non-empty
-        let top_piece = self.get(height).unwrap();
-        self.set(None, height);
-        return top_piece
-    }
-
-    /// Adds a piece to the top most position on the tower
-    /// Panics if the tower is full
-    pub fn drop_piece(&mut self, piece: Piece<'a>) {
-        let height = self.height();
-        use pieces::TowerHeight::*;
-        match height {
-            Top => panic!("Tower is full."),
-            Middle => self.set(Some(piece), Top),
-            Bottom => self.set(Some(piece), Middle),
-            Empty =>self.set(Some(piece), Bottom),
+    /// Returns a tower that has a piece added to the top most position on this tower
+    /// Returns Err if the tower is full does not modify Tower state when this happens
+    /// This function does not modify the original tower.
+    pub fn drop_piece<'b>(&'b self, piece: Piece<'a>) -> Result<Tower<'a>, &'static str> {
```

Comment:
I think this is equivalent to
```rust
pub fn drop_piece(&self, piece: Piece<'a>) -> Result<Tower<'a>, &'static str>
```
since the `'b` would be otherwise unconstrained